### ISSUE TITLE
paddle.vision.models中的模型怎么paddle.jit.save？

### ISSUE THREAD

**Issue by A-Pai** [PaddlePaddle/Paddle#31098]
https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/vision/models/lenet/LeNet_cn.html

如https://github.com/PaddlePaddle/Paddle/issues/31096
所讲：自定义的网络必须在forward前边添加@paddle.jit.to_static，否则后边的paddle.jit.save就会报错

直接使用paddle.vision.models中的已经实现好的经典模型，后边的paddle.jit.save也会报错
![image](https://user-images.githubusercontent.com/5112397/108652006-fde31e00-74fd-11eb-892a-fdf33514f748.png)

飞桨2.0的环境，实验代码为：

```import os
import random
import paddle
import numpy as np
from paddle.vision.models import LeNet
import paddle.nn.functional as F

model = LeNet()

def train(model):

    print('start training ... ')
    model.train()
    epoch_num = 1
    opt = paddle.optimizer.Momentum(learning_rate=0.001, momentum=0.9, parameters=model.parameters())

    # 使用Paddle自带的数据读取器
    train_loader = paddle.batch(paddle.dataset.mnist.train(), batch_size=10)
    valid_loader = paddle.batch(paddle.dataset.mnist.test(), batch_size=10)


    from visualdl import LogWriter
    with LogWriter(logdir="./log/train") as writer:
        for epoch in range(epoch_num):
            for batch_id, data in enumerate(train_loader()):
                # 调整输入数据形状和类型
                x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
                y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
                # 将numpy.ndarray转化成Tensor
                img = paddle.to_tensor(x_data)
                label = paddle.to_tensor(y_data)
                # 计算模型输出
                logits = model(img)
                # 计算损失函数
                loss = F.softmax_with_cross_entropy(logits, label)
                avg_loss = paddle.mean(loss)
                writer.add_scalar(tag="train_avg_loss", step=epoch*6000+batch_id+1, value=avg_loss)

                if batch_id % 1000 == 0:
                    print("epoch: {}, batch_id: {}, loss is: {}".format(epoch, batch_id, avg_loss.numpy()))
                avg_loss.backward()
                opt.step()
                opt.clear_grad()

            model.eval()
            accuracies = []
            losses = []
            for batch_id, data in enumerate(valid_loader()):
                # 调整输入数据形状和类型
                x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
                y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
                # 将numpy.ndarray转化成Tensor
                img = paddle.to_tensor(x_data)
                label = paddle.to_tensor(y_data)
                # 计算模型输出
                logits = model(img)
                pred = F.softmax(logits)
                # 计算损失函数
                loss = F.softmax_with_cross_entropy(logits, label)
                acc = paddle.metric.accuracy(pred, label)
                accuracies.append(acc.numpy())
                losses.append(loss.numpy())
            print("[validation] accuracy/loss: {}/{}".format(np.mean(accuracies), np.mean(losses)))
            model.train()

        # 保存模型参数
        paddle.save(model.state_dict(), 'mnist.pdparams')
# 创建模型
model = LeNet(num_classes=10)
# 启动训练过程
train(model)

path = "mmm"
paddle.jit.save(model, path)```

难道要手工打开已定义好的网络，在相应位置加“@paddle.jit.to_static”？

**Comment by lijianshe02** [comment]
paddle.jit.save适用于动转静之后的保存模型情况，您此时使用的经典模型是动态图模型，需要使用to_static装饰器作用转为静态图后才能使用这个接口保存模型

**Comment by A-Pai** [comment]
> paddle.jit.save适用于动转静之后的保存模型情况，您此时使用的经典模型是动态图模型，需要使用to_static装饰器作用转为静态图后才能使用这个接口保存模型

使用paddle.vision.models中的已经实现好的经典模型,怎样使用to_static装饰器？打开源代码？

**Comment by lijianshe02** [comment]
是的

**Comment by A-Pai** [comment]
> 是的

如果需要开膛破肚地去设置，就不好了，没有别的办法？
如果当前的版本没招儿，希望下个版本有招儿

**Comment by TCChenlong** [comment]
好的，我们会评估一下，感谢您的反馈~

**Comment by zhhsplendid** [comment]
嗨，现在paddle.jit.save可以支持第一个参数为Layer了。在您的样例里，请指定input_spec就可以了


> https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/vision/models/lenet/LeNet_cn.html
> 
> 如#31096
> 所讲：自定义的网络必须在forward前边添加@paddle.jit.to_static，否则后边的paddle.jit.save就会报错
> 
> 直接使用paddle.vision.models中的已经实现好的经典模型，后边的paddle.jit.save也会报错
> ![image](https://user-images.githubusercontent.com/5112397/108652006-fde31e00-74fd-11eb-892a-fdf33514f748.png)
> 
> 飞桨2.0的环境，实验代码为：
> 
> ```
> import random
> import paddle
> import numpy as np
> from paddle.vision.models import LeNet
> import paddle.nn.functional as F
> 
> model = LeNet()
> 
> def train(model):
> 
>     print('start training ... ')
>     model.train()
>     epoch_num = 1
>     opt = paddle.optimizer.Momentum(learning_rate=0.001, momentum=0.9, parameters=model.parameters())
> 
>     # 使用Paddle自带的数据读取器
>     train_loader = paddle.batch(paddle.dataset.mnist.train(), batch_size=10)
>     valid_loader = paddle.batch(paddle.dataset.mnist.test(), batch_size=10)
> 
> 
>     from visualdl import LogWriter
>     with LogWriter(logdir="./log/train") as writer:
>         for epoch in range(epoch_num):
>             for batch_id, data in enumerate(train_loader()):
>                 # 调整输入数据形状和类型
>                 x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
>                 y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
>                 # 将numpy.ndarray转化成Tensor
>                 img = paddle.to_tensor(x_data)
>                 label = paddle.to_tensor(y_data)
>                 # 计算模型输出
>                 logits = model(img)
>                 # 计算损失函数
>                 loss = F.softmax_with_cross_entropy(logits, label)
>                 avg_loss = paddle.mean(loss)
>                 writer.add_scalar(tag="train_avg_loss", step=epoch*6000+batch_id+1, value=avg_loss)
> 
>                 if batch_id % 1000 == 0:
>                     print("epoch: {}, batch_id: {}, loss is: {}".format(epoch, batch_id, avg_loss.numpy()))
>                 avg_loss.backward()
>                 opt.step()
>                 opt.clear_grad()
> 
>             model.eval()
>             accuracies = []
>             losses = []
>             for batch_id, data in enumerate(valid_loader()):
>                 # 调整输入数据形状和类型
>                 x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
>                 y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
>                 # 将numpy.ndarray转化成Tensor
>                 img = paddle.to_tensor(x_data)
>                 label = paddle.to_tensor(y_data)
>                 # 计算模型输出
>                 logits = model(img)
>                 pred = F.softmax(logits)
>                 # 计算损失函数
>                 loss = F.softmax_with_cross_entropy(logits, label)
>                 acc = paddle.metric.accuracy(pred, label)
>                 accuracies.append(acc.numpy())
>                 losses.append(loss.numpy())
>             print("[validation] accuracy/loss: {}/{}".format(np.mean(accuracies), np.mean(losses)))
>             model.train()
> 
>         # 保存模型参数
>         paddle.save(model.state_dict(), 'mnist.pdparams')
> # 创建模型
> model = LeNet(num_classes=10)
> # 启动训练过程
> train(model)
> 
> path = "mmm"
> paddle.jit.save(model, path)```
> 
> 难道要手工打开已定义好的网络，在相应位置加“@paddle.jit.to_static”？
> ```

**Comment by A-Pai** [comment]
> 嗨，现在paddle.jit.save可以支持第一个参数为Layer了。在您的样例里，请指定input_spec就可以了
> 
> > https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/vision/models/lenet/LeNet_cn.html
> > 如#31096
> > 所讲：自定义的网络必须在forward前边添加@paddle.jit.to_static，否则后边的paddle.jit.save就会报错
> > 直接使用paddle.vision.models中的已经实现好的经典模型，后边的paddle.jit.save也会报错
> > ![image](https://user-images.githubusercontent.com/5112397/108652006-fde31e00-74fd-11eb-892a-fdf33514f748.png)
> > 飞桨2.0的环境，实验代码为：
> > ```
> > import random
> > import paddle
> > import numpy as np
> > from paddle.vision.models import LeNet
> > import paddle.nn.functional as F
> > 
> > model = LeNet()
> > 
> > def train(model):
> > 
> >     print('start training ... ')
> >     model.train()
> >     epoch_num = 1
> >     opt = paddle.optimizer.Momentum(learning_rate=0.001, momentum=0.9, parameters=model.parameters())
> > 
> >     # 使用Paddle自带的数据读取器
> >     train_loader = paddle.batch(paddle.dataset.mnist.train(), batch_size=10)
> >     valid_loader = paddle.batch(paddle.dataset.mnist.test(), batch_size=10)
> > 
> > 
> >     from visualdl import LogWriter
> >     with LogWriter(logdir="./log/train") as writer:
> >         for epoch in range(epoch_num):
> >             for batch_id, data in enumerate(train_loader()):
> >                 # 调整输入数据形状和类型
> >                 x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
> >                 y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
> >                 # 将numpy.ndarray转化成Tensor
> >                 img = paddle.to_tensor(x_data)
> >                 label = paddle.to_tensor(y_data)
> >                 # 计算模型输出
> >                 logits = model(img)
> >                 # 计算损失函数
> >                 loss = F.softmax_with_cross_entropy(logits, label)
> >                 avg_loss = paddle.mean(loss)
> >                 writer.add_scalar(tag="train_avg_loss", step=epoch*6000+batch_id+1, value=avg_loss)
> > 
> >                 if batch_id % 1000 == 0:
> >                     print("epoch: {}, batch_id: {}, loss is: {}".format(epoch, batch_id, avg_loss.numpy()))
> >                 avg_loss.backward()
> >                 opt.step()
> >                 opt.clear_grad()
> > 
> >             model.eval()
> >             accuracies = []
> >             losses = []
> >             for batch_id, data in enumerate(valid_loader()):
> >                 # 调整输入数据形状和类型
> >                 x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
> >                 y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
> >                 # 将numpy.ndarray转化成Tensor
> >                 img = paddle.to_tensor(x_data)
> >                 label = paddle.to_tensor(y_data)
> >                 # 计算模型输出
> >                 logits = model(img)
> >                 pred = F.softmax(logits)
> >                 # 计算损失函数
> >                 loss = F.softmax_with_cross_entropy(logits, label)
> >                 acc = paddle.metric.accuracy(pred, label)
> >                 accuracies.append(acc.numpy())
> >                 losses.append(loss.numpy())
> >             print("[validation] accuracy/loss: {}/{}".format(np.mean(accuracies), np.mean(losses)))
> >             model.train()
> > 
> >         # 保存模型参数
> >         paddle.save(model.state_dict(), 'mnist.pdparams')
> > # 创建模型
> > model = LeNet(num_classes=10)
> > # 启动训练过程
> > train(model)
> > 
> > path = "mmm"
> > paddle.jit.save(model, path)```
> > 
> > 难道要手工打开已定义好的网络，在相应位置加“@paddle.jit.to_static”？
> > ```

可否给个完整的例子？

**Comment by zhhsplendid** [comment]
关键代码类似：
model = LeNet(num_classes=10)
input_var = paddle.zeros(shape=xxx) # 指定您需要的输入shape
paddle.jit.save(model, path, input_spec=[input_var])

具体完整的例子见：
https://github.com/PaddlePaddle/FluidDoc/blob/develop/doc/paddle/guides/04_dygraph_to_static/basic_usage_cn.rst

里面这段代码：
```
import paddle

fc_layer = SimpleFcLayer(3, 4, 2)
in_np = np.random.random([3, 4]).astype('float32')
input_var = paddle.to_tensor(in_np)
out = fc_layer(input_var)

paddle.jit.save(fc_layer, "./fc_layer_dy2stat", input_spec=[input_var])
```

**Comment by A-Pai** [comment]
> 关键代码类似：
> model = LeNet(num_classes=10)
> input_var = paddle.zeros(shape=xxx) # 指定您需要的输入shape
> paddle.jit.save(model, path, input_spec=[input_var])
> 
> 具体完整的例子见：
> https://github.com/PaddlePaddle/FluidDoc/blob/develop/doc/paddle/guides/04_dygraph_to_static/basic_usage_cn.rst
> 
> 里面这段代码：
> 
> ```
> import paddle
> 
> fc_layer = SimpleFcLayer(3, 4, 2)
> in_np = np.random.random([3, 4]).astype('float32')
> input_var = paddle.to_tensor(in_np)
> out = fc_layer(input_var)
> 
> paddle.jit.save(fc_layer, "./fc_layer_dy2stat", input_spec=[input_var])
> ```

还是需要开膛破肚地到网络定义的地方加上“@paddle.jit.to_static”，如果使用的是框架提供的已有网络，这样做好吗？
![QQ截图20210428121655](https://user-images.githubusercontent.com/5112397/116346005-dec57d00-a81b-11eb-9afc-ea2b7cdd5463.jpg)
